Add support for custom trained PunktTokenizer in PreProcessor

danielbichuetti opened the pull request:
**Related Issue(s)**: #2780 #2781

**Proposed changes**:
- Today, the PreProcessor makes usages of _NLTK PunktTokenizer_. On some specific domains the default model demands some extra training on a corpus linked to the domain. Based on this, I propose to introduce a parameter `tokenizer_model_folder` that allows users to use custom trained models. The naming would be something like `language.pickle`. This change doesn't break current class and doesn't make impossible to use same parameter for another Sentence Tokenizer if on the future project decides for it (Spacy, Stanza as examples).
 If for that specific language, a model is present on this folder, PreProcessor would use it. If not, fallback to default one. If anyone wants he could have a folder with domains linked to law, another for medical tokenizer and so on. 

This is the first draft on the PR, so I can know where it can be further improved.

## Pre-flight checklist
- [x]  I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md)
- [x] I have [enabled actions on my fork](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md#forks)
- [x] If this is a code change, I added tests or updated existing ones 
- [x] If this is a code change, I updated the docstrings

ZanSara reviewed (commented):
Hey @danielbichuetti! Nice one! :+1: I added some comments on the usage of `pathlib`, a note on the tests and some other minor comments, but there's no issue with your approach. PR looks already very good as it is! Let me know if anything is unclear.

danielbichuetti commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -95,7 +98,9 @@ def __init__(
         self.split_respect_sentence_boundary = split_respect_sentence_boundary
-        self.language = iso639_to_nltk.get(language, language)
+        self.tokenizer_model_folder = tokenizer_model_folder
+        self.language = language
+        self.language_name = iso639_to_nltk.get(self.language, self.language)
```
I was trying to avoid having to convert it back again. 

The custom model would be stored like `en.pickle.` This way, chances of bad typo (user saving model with a slightly different name than expected) would be minimized. Since it's an optional parameter and on other methods it can be `None`, we would not be able to get data back from Dict without doing some dictionary search. Do you think the dict search approach is better than an extra variable ?

danielbichuetti commented:
@ZanSara I just read another topic regarding usage of Catalan on the sentence tokenizer, which is not a default language for Punkt. The way this custom model folder is designed (per your request about removing that extra variable) it allows custom models for languages that NLTK already support. And the name of the file is taken from the language parameter, which is translated based on a list that is defined on code. Don't you think this is a good opportunity to use the ISO ? 

One example:

1) User sets language as Catalan.
2) User sets custom folder as "models"
3) If we keep using the long name, user would need to manually edit code and add on dictionary. And then he will be able to use Catalan.

On the other side:

3) Code allows the default dict and set long name. We set another variable with ISO. When loading, we try to get this file and if not we fall back to the default language (code now would default to english).

Don't you think this is a good example of why keeping a separate ISO variable ? Or maybe any other idea that made this more general.

I'm not quite sure If I managed to translate my toughts in a good way. :smiley:

danielbichuetti commented:
@ZanSara  I have refactored the code. Due to the fallback auto-handling adopted, code was separated and logs are specific for each scenario. This way, users can find his own errors easily.

I added some extra tests, following your suggestions. The small Portuguese PunktSentenceTokenizer model on samples folder has achieved results on legal documents way higher than default ones.

danielbichuetti commented:
What sounds interesting is that on the GH Workflow test is failing because it can't determine the format of the model `pt.pickle`:

`ERROR    haystack.nodes.preprocessor.preprocessor:preprocessor.py:484 PreProcessor couldn't determine model format of sentence tokenizer at /home/runner/work/haystack/haystack/test/samples/preprocessor/nltk_models/pt.pickle.`

On VS Code and command-line on local machine, the pt.pickle is correct. Yesterday I uploaded a huge model which was not for legal. Then I committed this small for the current text. Does workflow has some caching enabled ? Something that may be loading the old pickle model ?

ZanSara commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +457,49 @@ def _find_longest_common_ngram(
+                logger.error(
+                    f"PreProcessor couldn't determine model format of sentence tokenizer at {str(tokenizer_model_path)}."
+                )
+            if len(sentences) == 0 and language_name is None:
```
Just a minor thing here: `language_name is not None` will be `True` if `language_name == ""`. With strings I think it's safer to check with `not language_name`, which will evaluate to `False` if `language_name == ""`. I know that in the current code this can't happen, but I find it a bit more future-proof.

Similarly, you can check for empty lists with `not sentences`.

```suggestion
            if not sentences and not language_name:
```

danielbichuetti replied:
Old habits from C#, checking for null or empty :smiling_face_with_tear:

ZanSara commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +457,49 @@ def _find_longest_common_ngram(
+                    f"PreProcessor couldn't find default or custom sentence tokenizer model for {language}. Using English instead."
+                )
+                sentences = nltk.tokenize.sent_tokenize(text, language="english")
+            elif len(sentences) == 0 and language_name is not None:
```
Same as above
```suggestion
            elif not sentences and language_name:
```

ZanSara commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +457,49 @@ def _find_longest_common_ngram(
+                    f"PreProcessor couldn't find custom sentence tokenizer model for {language}. Using default {language} model."
+                )
+                sentences = nltk.tokenize.sent_tokenize(text, language=language_name)
+        elif language_name is None:
```
Same as above
```suggestion
        elif not language_name:
```

ZanSara commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +457,49 @@ def _find_longest_common_ngram(
+        sentences = []
+        language_name = iso639_to_nltk.get(language)
+        tokenizer_model_path = (
+            Path(tokenizer_model_folder) / f"{language}.pickle" if tokenizer_model_folder is not None else None
```
What if `language` is `None` here? Do you think we can give a default, or is better to check if it's `None` before building this path?

In general I think the whole `if-else` chain can be simplified a bit if we check `if language:` early on.

danielbichuetti replied:
If `language` is `None`, `.get()` on the dict will return `None`.

Oh, I didn't set a default value at that point to avoid the usage of another variable, so we could log the different “behaviors” of the fallback:

1) Use default model for language X

2) Use custom model for language X
2.1) There is a default model for this language X in case of custom model lookup/format error.
2.2) There is no default model for this language X in case of error, and tokenizer fallback to English

Currently, we're logging all the possible scenarios with different messages.

```python
if not sentences and not language_name:
                logger.error(
                    f"PreProcessor couldn't find default or custom sentence tokenizer model for {self.language}. Using English instead."
                )
                sentences = nltk.tokenize.sent_tokenize(text, language="english")
elif not sentences and language_name:
                logger.error(
                    f"PreProcessor couldn't find custom sentence tokenizer model for {self.language}. Using default {self.language} model."
                )
                sentences = nltk.tokenize.sent_tokenize(text, language=language_name)
```

This part and the below it could be simplified with usage of `flags`, but the `linter` doesn't like it. If we want to keep the different error logs, we will need in flags and more ifs again. What do you think ? The more verbose error log could make easier the life of users (usually there is some sort of despair when they find errors).

ZanSara reviewed (commented):
Ok, I changed my mind a bit and I have a new proposal for `tokenizer_model_folder`. Nothing big, but I want to know what do you think about this :slightly_smiling_face: I also have a few comments on the new method, but again it's mostly small technicalities.

danielbichuetti commented on haystack/nodes/preprocessor/base.py:
```diff
@@ -20,6 +22,7 @@ def process(
         split_length: Optional[int] = 1000,
         split_overlap: Optional[int] = None,
         split_respect_sentence_boundary: Optional[bool] = True,
+        tokenizer_model_folder: Optional[Union[str, Path]] = None,
```
Since the tokenizer is extremely connected with the language, that will be really better!

danielbichuetti commented:
> Super interesting, might be a bug in the way we do the caching. Share the findings if you have any, I'm going to investigate 

Yeah, I got a headache about this issue.

The issue is being caused because I saved using protocol version 5. As you can see from Python documentation:

> Protocol version 5 was added in Python 3.8. It adds support for out-of-band data and speedup for in-band data. Refer to [PEP 574](https://www.python.org/dev/peps/pep-0574) for information about improvements brought by protocol 5.

Repo workflow is running using Python 3.7, so the highest allowed protocol version is 4.

I'll generate a model using version 4 :smile:

ZanSara commented:
Thank you! Sounds like a good idea. Haystack supports Python versions back to 3.7, so we should make sure that's always the case :slightly_smiling_face:

One more thing: the failure you see in `integration-tests-linux (nodes)` and `integration-tests-windows (nodes)` was fixed in master this morning. Please merge master so the CI becomes green again.

ZanSara commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +450,49 @@ def _find_longest_common_ngram(
+        :return: list[str], list of sentences
+        """
+        sentences = []
+        language_name = iso639_to_nltk.get(self.language)
+        tokenizer_model_path = (
+            Path(self.tokenizer_model_folder).absolute() / f"{self.language}.pickle"
+            if self.tokenizer_model_folder is not None
+            else None
+        )
+        if tokenizer_model_path is not None:
+            try:
+                sentence_tokenizer = nltk.data.load(f"file:{str(tokenizer_model_path)}", format="pickle")
+                sentences = sentence_tokenizer.tokenize(text)
+            except LookupError:
+                logger.warning(f"PreProcessor couldn't load sentence tokenizer from {str(tokenizer_model_path)}")
+            except (UnpicklingError, ValueError) as e:
+                logger.warning(
+                    f"PreProcessor couldn't determine model format of sentence tokenizer at {str(tokenizer_model_path)}. - Exception: {str(e)}"
+                )
+            if not sentences and not language_name:
+                logger.warning(
+                    f"PreProcessor couldn't find default or custom sentence tokenizer model for {self.language}. Using English instead."
+                )
+                sentences = nltk.tokenize.sent_tokenize(text, language="english")
+            elif not sentences and language_name:
+                logger.warning(
+                    f"PreProcessor couldn't find custom sentence tokenizer model for {self.language}. Using default {self.language} model."
+                )
+                sentences = nltk.tokenize.sent_tokenize(text, language=language_name)
+        elif not language_name:
+            logger.warning(
+                f"PreProcessor couldn't find default sentence tokenizer model for {self.language}. Using English instead. \
+                    You may train your own model and use the tokenizer_model_folder parameter."
+            )
+            sentences = nltk.tokenize.sent_tokenize(text, language="english")
+        else:
+            sentences = nltk.tokenize.sent_tokenize(text, language=language_name)
+
```
Here is my take on the same code. By using many `return` I reduce the number of `if-else`. I added a few comments and note that I raised all log messages to `ERROR`: I think they're quite important, let's make the user aware.

There should be no difference with your version. No need to choose between these two, feel free to mix them :slightly_smiling_face: 

```suggestion

        language_name = iso639_to_nltk.get(self.language)
        
        # Try to load a custom model from 'tokenizer_model_path'
        if self.tokenizer_model_folder:
            tokenizer_model_path = Path(self.tokenizer_model_folder).absolute() / f"{self.language}.pickle"
            try:
                sentence_tokenizer = nltk.data.load(f"file:{str(tokenizer_model_path)}", format="pickle")
                sentences = sentence_tokenizer.tokenize(text)
            except LookupError:
                logger.exception(f"PreProcessor couldn't load sentence tokenizer from {str(tokenizer_model_path)}")
            except (UnpicklingError, ValueError) as e:
                logger.exception(
                    f"PreProcessor couldn't determine model format of sentence tokenizer at {str(tokenizer_model_path)}."
                )
            if sentences:
                return sentences
            
            # NLTK failed to split, fallback to the default model or to English
            if language_name:
               logger.error(
                    f"PreProcessor couldn't find custom sentence tokenizer model for {self.language}. Using default {self.language} model."
                )
                return nltk.tokenize.sent_tokenize(text, language=language_name)
             logger.error(
                f"PreProcessor couldn't find default or custom sentence tokenizer model for {self.language}. Using English instead."
            )
            return nltk.tokenize.sent_tokenize(text, language="english")
        
        # Use a default NLTK model
        if language_name:
            return nltk.tokenize.sent_tokenize(text, language=language_name)
        logger.error(
            f"PreProcessor couldn't find default sentence tokenizer model for {self.language}. Using English instead. "
            "You may train your own model and use the 'tokenizer_model_folder' parameter."
        )
        return nltk.tokenize.sent_tokenize(text, language="english")
```

danielbichuetti replied:
:smiling_face_with_tear: What a shame! 

When I did the flag tries, there wasn't this function. Code was inside another method, so the returns were not possible. After the method creation I didn't took a look and kept that in mind lol.

This is what I can call a silly mistake.

danielbichuetti commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -232,7 +237,7 @@ def clean(
         if clean_header_footer:
             text = self._find_and_remove_header_footer(
-                text, n_chars=300, n_first_pages_to_ignore=1, n_last_pages_to_ignore=1
+                str(text), n_chars=300, n_first_pages_to_ignore=1, n_last_pages_to_ignore=1
```
Great, I'll double-check the text variable. I tried to minimize code changes on this. I just noticed the string and binary issues because of the linter warns. Then I noticed it could be string or DataFrame. I'll include pre-check on text.

tstadel commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +450,49 @@ def _find_longest_common_ngram(
             longest = ""
         return longest if longest.strip() else None
+
+    def _sentence_tokenizer(self, text: str) -> List[str]:
```
Could we please rename that method to a verb-form emphasizing the action is being executed like`_split_sentences`? This would make it much clearer what to expect from that method. Currently it looks like passing some context (the text) to an object that at some later point might work on that context.

tstadel reviewed (changes requested):
I added a small comment about renaming the method as it currently doesn't feel intuitive to me. Besides that looks cool.

danielbichuetti commented:
CI is red because of timeout on GitHub Action (30 min).

danielbichuetti commented:
@tstadel I have refactored the method name as per your request. Much more intuitive now :smiley: 

@ZanSara The improvements you proposed have been committed. :smile:

ZanSara reviewed (approved):
Looks good! Thank you :slightly_smiling_face:

tstadel commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +459,52 @@ def _find_longest_common_ngram(
+        """
+        Tokenize text into sentences.
+        :param text: str, text to tokenize
+        :param language: str, language of text
+        :param tokenizer_model_path: Path, path to tokenizer model
```
These params are superfluous here.

tstadel commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +459,52 @@ def _find_longest_common_ngram(
+        language_name = iso639_to_nltk.get(self.language)
+
+        # Try to load a custom model from 'tokenizer_model_path'
+        if self.tokenizer_model_folder:
```
Let's make the check for None explicit.

```suggestion
        if self.tokenizer_model_folder is not None:
```

tstadel commented on haystack/nodes/preprocessor/preprocessor.py:
```diff
@@ -444,3 +459,52 @@ def _find_longest_common_ngram(
+            return nltk.tokenize.sent_tokenize(text, language="english")
+
+        # Use a default NLTK model
+        if language_name:
```
And here too:
```suggestion
        if language_name is not None:
```

danielbichuetti replied:
Firstly, I have done this way, but @ZanSara have made this point:

> Just a minor thing here: language_name is not None will be True if language_name == "". With strings I think it's safer to check with not language_name, which will evaluate to False if language_name == "". I know that in the current code this can't happen, but I find it a bit more future-proof.

I have committed the `if x is not None` version again, please advise of any further changes,

tstadel replied:
Sorry for bringing this up again, I think an explicit None check is just more clear about what we're checking here for. Without it it's kind of harder to understand what happens if someone passes `""` or even `[]` (especially for the model path). I would find it unintuitive for this method to just work as if the params hasn't been set in that cases.

tstadel reviewed (changes requested):
@danielbichuetti I found some minor things. However they are worth fixing before merging. Could you please remove the superfluous params within the docstring and make the check for None explicit?

danielbichuetti commented:
@tstadel Hi, may you review the changes? @ZanSara approved, waiting your review now.
